Added restaurant menu

bauke2012 opened the pull request:
Added CERN restaurant menu from Novae API
Usage
config:

# Restaurant

# enable CERN restaurant menu display
menuEnabled=1
# select restaurant number (1,2,3)
restaurant=3
# At what hour to switch from today's to tomorrow's menu (somewhere after lunchtime)
menuSwitchHour=14 #

bartoszbielawski commented:
Hi, 

Thanks for the contribution. From what I've already seen the menu will be displayed whole day long. Could you limit it to be displayed only between 10 and 14 (configurable)? 

I've seen some other small issues (like sleeping instead of yielding), I will try to review the code soon. 

Cheers,
Bartosz

bauke2012 commented:
Hi Bartosz,

Thank you. Could not get this idea out of my head so might as well contribute.

Currently it will switch to display tomorrow's menu from set menuSwitchHour. eg. 14 = after 14:00
I can change this to configurable start and end time for displaying today's menu, with a config option to show tomorrow's menu after the end time (default no). Would that be okay?

I see the delay is left in the HTTP error retry loop from crude testing, will change that. At the end of run, I use sleep from tasks.hpp similar to other tasks. Is this the correct use, or should I modify it?

Looking forward to your feedback.
Bauke

bartoszbielawski commented:
> Thank you. Could not get this idea out of my head so might as well contribute.

I did not occur to me that there's a web API for the menus, but I have to admit many people in my corridor will be thrilled be the new feature.

I would keep it simple. Define start and stop times of the display and then hardcode the switch time to be 14:00. Don't think serving times will change.

> I see the delay is left in the HTTP error retry loop from crude testing, will change that.

This was one of the things I've found.

>  I use sleep from tasks.hpp similar to other tasks. Is this the correct use, or should I modify it?

That's the correct way to handle it. I never wondered how long the sleep can be, I need to check ;) But the task could sleep till another day.

The code for fetching HTTP is one level below the task method, so I guess you will need to return something so that the method just sleeps there and returns.

I've seen you have used another JSON parses. I had issues with some of these in the past due to memory consumption, that's why there's the custom one. Not sure you want to change your code - if not - I understand.

BB

bauke2012 commented:
Hi Bartosz,

> I did not occur to me that there's a web API for the menus, but I have to admit many people in my corridor will be thrilled be the new feature.

Looking forward to their reactions.

> I would keep it simple. Define start and stop times of the display and then hardcode the switch time to be 14:00. Don't think serving times will change.

I have implemented config definitions for start and stop time (default 7, 14), and optional display of tomorrow's menu after the stop time (default no). Modify as you wish.

Whenever no menu is available (e.g. weekend) or time is outside configured, the message will respond empty. Confirmed that this skips displaying the message.

> > I see the delay is left in the HTTP error retry loop from crude testing, will change that.

Changed to sleep

> That's the correct way to handle it. I never wondered how long the sleep can be, I need to check ;) But the task could sleep till another day.

I have left sleep time 900_s. I noticed French menu is available days ahead, but the English version of the menu is added later at arbitrary time. (1-3 hours before lunch). EN is prioritized over FR menu.

> I've seen you have used another JSON parses. I had issues with some of these in the past due to memory consumption, that's why there's the custom one. Not sure you want to change your code - if not - I understand.

Gave this a try, but left it to be for now as I'm limited on time.


An addition I did is a retry loop for the temperature sensor as ours seems to fail quite often. Feel free to leave out or merge as you like.

Cheers,
Bauke
